AVM Boxes Ops in Pyteal (#438)
algorand/pyteal: docs/state.rst

> State Operation Table

Author: just insert a table to explain allowed types. let me know if it looks right.

> State Operation Table

Reviewer: @ahangsu Resolving - Table looks good.  Thanks for bringing in the feedback.

---------------------

================== ======================= ======================== ======================== ===================== ======================
Context            Create                  Write                    Read                     Delete                Check If Exists
================== ======================= ======================== ======================== ===================== ======================
Current App Global                         :any:`App.globalPut`     :any:`App.globalGet`     :any:`App.globalDel`  :any:`App.globalGetEx`
Current App Local                          :any:`App.localPut`      :any:`App.localGet`      :any:`App.localDel`   :any:`App.localGetEx`
Other App Global                                                    :any:`App.globalGetEx`                         :any:`App.globalGetEx`
Other App Local                                                     :any:`App.localGetEx`                          :any:`App.localGetEx`

> Current App Boxes  | :any:`App.box_create` | :any:`App.box_put`     | :any:`App.box_extract` :any:`App.box_delete` :any:`App.box_length`

Author: yes, by nature `App.box_get` can be used to check box existence, for it returns 2 values on stack, one of them is box existence.

        otherAppOtherAccountRole,
        If(otherAppOtherAccountRole.hasValue(), otherAppOtherAccountRole.value(), Bytes("none"))
    ])

Box Storage
-----------

Box storage consists of key-value pairs that are stored in application's local context.

The app account's minimum balance requirement (MBR) is increased by each box, and by the byte size of the box.

> If one deletes an application with outstanding boxes, the MBR is not recoverable from the deleted app account.

Reviewer 2: ```suggestion
WARNING: If one deletes an application with outstanding boxes, the MBR is not recoverable from the deleted app account. It is recommended that _before_ app deletion, all box storage be deleted, and funds previously allocated to the MBR be withdrawn.
```

Box sizes and names cannot be changed after initial allocation, but they can be deleted and re-allocated.

> Boxes are only visible to the app itself, other apps cannot view the app's own boxes on-chain.

Reviewer 2: ```suggestion
Boxes are only visible to the app itself; in other words, an app cannot read from or write to another app's boxes on-chain.
```

Author: it just reminds me of my advisor and I are working through paper... he's picking verbal error from each line of my writing... quite a nostalgic feeling, but probably quite painful for my advisor

Boxes can be manipulated as follows:

Create Boxes
~~~~~~~~~~~~

To create a box, use :any:`App.box_create`, or :any:`App.box_put` method.

For :any:`App.box_create`, the first argument is the box name, and the second argument is the byte size to be allocated.

For example:

.. code-block:: python

    # Allocate a box called "BoxForAlice" of byte size 100

>     App.box_create(Bytes("BoxForAlice"), Int(0xA11CE))

Author: oh boy, hex range gets wild. I will trim the number down

    # Allocate a box called "BoxForBob" of byte size 90

>     App.box_create(Bytes("BoxForBob"), Int(0xB0B)

Reviewer 2: Similar here, was the intention to just use the commented sizes?

Reviewer 2: (and a missing right paren)

Author: I just changed the example code block to match with what comments suggested.

For :any:`App.box_put`, the first argument is the box name to create or to write to, and the second argument is the bytes to write.

> **Notice**: if the box exists, then :any:`App.box_put` will write the contents to the box;

Reviewer 2: ```suggestion
**Notice**: if the box exists, then :any:`App.box_put` will write the contents to the box (assuming the replacement has length identical to the box'es size, as otherwise the operation will fail);
```

otherwise, it will create a box containing exactly the same input bytes.

.. code-block:: python

    # create a 42 bytes length box called `poemLine` with content
    App.box_put(Bytes("poemLine"), Bytes("Of that colossal wreck, boundless and bare"))
    # write to box `poemLine` with new value

>     App.box_put(Bytes("poemLine"), Bytes("The lone and level sands stretch far away."))

Reviewer 2: very creative!

Writing to a Box
~~~~~~~~~~~~~~~~

To write to a box, use :any:`App.box_replace`, or :any:`App.box_put` method.

> For :any:`App.box_replace`, it writes bytes of certain length from a start index in a Box.

Reviewer 2: ```suggestion
:any:`App.box_replace` writes bytes of certain length from a start index in a Box.
```

The first argument is the box name to write into, the second argument is the starting index to write,
and the third argument is the replacement bytes. For example:

.. code-block:: python

>    # replace 2 bytes to `Ne` starting from 0st bytes in a box with name `wordleBox`

Reviewer 2: ```suggestion
   # replace in the box named `wordleBox` by the 2 bytes `Ne` starting from the 0'th byte
```

   App.box_replace(Bytes("wordleBox"), Int(0), Bytes("Ne"))

> For :any:`App.box_put`, it writes the full content to a pre-existing box.

Reviewer 2: ```suggestion
:any:`App.box_put` writes the full contents to a pre-existing box.
```

Reviewer 2: nit: maybe this section is a bit redundant given the example for `box_put` previously provided? But it's also arguable to keep this example given that this section is about "writing". A middle ground would be to just mention that ":any:`App.box_put` can also be used for writing a box'es entire contents, as seen in the example above."

Author: indeed, it is redundant for `box_put`, as it was previously stated. I was thinking of a way to circumvent re-writing same paragraph cross different sections.

The middle ground suggestion makes sense to me, I will adapt to that suggestion and reflect in the document.

   App.box_put(Bytes("twelveByteBox"), Bytes("fullContents"))

Reading from a Box
~~~~~~~~~~~~~~~~~~

To read from a box, use :any:`App.box_extract`, or :any:`App.box_get` method.

For :any:`App.box_extract`, it reads bytes of a certain length from a start index in a Box.
The first argument is the box name to read from, the second argument is the starting index to read,
and the third argument is the length of bytes to extract. For example:

.. code-block:: python

   # extract segment of length 10 from 5th byte in Box of name `NoteBook`
   App.box_extract(Bytes("NoteBook"), Int(5), Int(10))

For :any:`App.box_get`, it gets the full content of a Box.
The only argument is the box name, and it returns a :any:`MaybeValue` containing:

- a boolean value indicating if the box exists
- the full content of the box.

For example:

.. code-block:: python

   # get the full content from a Box of name `NoteBook`
   App.box_get(Bytes("NoteBook"))

Deleting a Box
~~~~~~~~~~~~~~

To delete a box, use :any:`App.box_delete` method. The only argument is the box name. For example:

.. code-block:: python

    App.box_delete(Bytes("boxToRemove"))

> Checking if a Box Exists

Author: 👍

        otherAppOtherAccountRole,
        If(otherAppOtherAccountRole.hasValue(), otherAppOtherAccountRole.value(), Bytes("none"))
    ])

Box Storage
-----------

Box storage consists of key-value pairs that are stored in an application's local context.

> The app account's minimum balance requirement (MBR) is increased with each additional box, and each additional byte in the box'es name and allocated size.

Author: 👍

.. warning::

   If one deletes an application with outstanding boxes, the MBR is not recoverable from the deleted app account.
   It is recommended that *before* app deletion, all box storage be deleted, and funds previously allocated to the MBR be withdrawn.

Box sizes and names cannot be changed after initial allocation, but they can be deleted and re-allocated.
Boxes are only visible to the application itself; in other words, an application cannot read from or write to another application's boxes on-chain.

> Boxes can be manipulated as follows:

Reviewer: @ahangsu Minor and optional:  I'd expect an immediate listing following a colon (:) rather than a series of subsections.

I don't feel strongly and you're welcomed to resolve as is.  Alternatives include (a) removing the line or (b) rephrasing (_suggestion_ offers one phrasing).

```suggestion
The following sections explain how to work with boxes.
```

Author: I would like to rephrase this as you suggested, feels making more sense.

> .. _Create Boxes:
> 
> Create Boxes

Reviewer: @ahangsu Minor:  For consistency with the sections' headers, I propose a minor tweak.

```suggestion
.. _Creating Boxes:

Creating Boxes
```

Author: yes, I will take care of the section replacing.

~~~~~~~~~~~~

To create a box, use :any:`App.box_create`, or :any:`App.box_put` method.

:any:`App.box_create` makes a box with a specified name and byte length.
The first argument is the box name, and the second argument is the byte size to be allocated.

For example:

.. code-block:: python

    # Allocate a box called "BoxA" of byte size 100
    App.box_create(Bytes("BoxA"), Int(100))
    # Allocate a box called "BoxB" of byte size 90
    App.box_create(Bytes("BoxB"), Int(90)

For :any:`App.box_put`, the first argument is the box name to create or to write to, and the second argument is the bytes to write.

.. note::

   If the box exists, then :any:`App.box_put` will write the contents to the box

>    (assuming the replacement has length **identical** to the box'es byte size, as otherwise the operation will fail);

Reviewer: @ahangsu Alternatively and optionally, the phrasing can be made more concise:

If the box exists, then App.box_put will write the contents to the box (fails when the replacement length is not identical to the box's byte size);

>    (assuming the replacement has length **identical** to the box'es byte size, as otherwise the operation will fail);

Author: I go with the more concise version of rephrasing.

- a boolean value indicating if the box exists
- the full contents of the box.

For example:

.. code-block:: python

   # get the full contents from a box named `NoteBook`
   App.box_get(Bytes("NoteBook")).value()

.. note::

   :any:`App.box_get` can also be used to check the existence of a box. For example:

   .. code-block:: python

      # check existence of a box named `NoteBook`
      App.box_get(Bytes("NoteBook")).hasValue()

Deleting a Box
~~~~~~~~~~~~~~

To delete a box, use :any:`App.box_delete` method. The only argument is the box name. For example:

.. code-block:: python

    App.box_delete(Bytes("boxToRemove"))

Checking if a Box Exists and Reads its Length
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To check the existence of a box, use the :any:`App.box_length` method.
The only argument is the box name, and it returns a :any:`MaybeValue` containing:

- a boolean value indicating if the box exists
- the actual byte size of the box.

For example:

.. code-block:: python

   # search for the box length for box `someBox`, and get the bool value for box existence
   App.box_length(Bytes("someBox")).hasValue()

.. note::

>    :any:`App.box_get` can also check the existence of a box, mentioned in `Reading from a Box`_.

Reviewer 2: ```suggestion
   :any:`App.box_get` can also check the existence of a box as mentioned in `Reading from a Box`_.
```

Author: 👍
